### PR TITLE
Add `IpAddr` <-> `ipaddress.IPv(4/6)Address` conversion

### PR DESCRIPTION
**Pull request by mhils**
This PR adds support for conversion between Python's `ipaddress.IPv4Address`/`ipaddress.IPv6Address` and `std::net::IpAddr`. :smiley:

### CODE REVIEW

File: src/conversions/std/ipaddr.rs
Comment (mhils) [inline src/conversions/std/ipaddr.rs:9]: There are a few occasions where this particular GILOnceCell pattern is used (e.g. [`types/mapping.rs`](https://github.com/PyO3/pyo3/blob/451729aef0e787a85011dc608433dea9c663ecee/src/types/mapping.rs#L118), [`conversions/rust_decimal.rs`](https://github.com/PyO3/pyo3/blob/451729aef0e787a85011dc608433dea9c663ecee/src/conversions/rust_decimal.rs#L76)). Happy to move PyClassRef outside of this module and use it consistently around the codebase if you feel that'd be an improvement. :)

Comment (mhils) [reply src/conversions/std/ipaddr.rs]: Oh, that's neat! Implemented in https://github.com/PyO3/pyo3/pull/3197/commits/721c13ce30782415606c5a06a934378eace24415. If y'all feel that's generally useful I'd be happy to apply it outside of the module. :)

File: src/conversions/std/ipaddr.rs
Comment (adamreichold) [inline src/conversions/std/ipaddr.rs:20]: I think those statics could have their scope reduced by moving them into the `to_object` bodies.

File: src/conversions/std/ipaddr.rs
Comment (mhils) [reply src/conversions/std/ipaddr.rs]: Probably easiest to centralize that in a follow-up PR, any objections to that? :)

**Comment by mhils** [comment]
> (Please note that you'll either need to squash or rebase into a coherent history at the end before meging.)

To clarify, does this mean that your automation does not squash automatically an I am expected squash myself manually? Happy to do that, I didn't find anything specific in `Contributing.md` and expected you'd prefer to just press GitHub's squash merge button in the end. :)

**Comment by DataTriny** [comment]
Hello @mhils 

Thank you for contributing to PyO3!

The project is undergoing changes on its licensing model to better align with the rest of the Rust crates ecosystem.

Before your work gets merged, please express your consent by leaving a comment on this [pull request](https://github.com/PyO3/pyo3/pull/3108).

Thank you!

**Comment by adamreichold** [comment]
> > (Please note that you'll either need to squash or rebase into a coherent history at the end before meging.)
> 
> To clarify, does this mean that your automation does not squash automatically an I am expected squash myself manually? Happy to do that, I didn't find anything specific in `Contributing.md` and expected you'd prefer to just press GitHub's squash merge button in the end. :)

Our automation is currently bors and it can only always merge or always squash. Since we do like coherent multi-commit histories, we do not squash and hence you need to either squash manually or provide a coherent multi-commit history.

(This is not currently documented, but it is probably not useful ATM as bors-ng will not available for long and we will need to switch to GitHub's merge queues (which seem to have the same issue) soon.)

File: src/conversions/std/ipaddr.rs
Comment (adamreichold) [reply src/conversions/std/ipaddr.rs]: Please still move the method definition to `src/sync.rs` even it is only used here (the visibility is not affected by this in any case).

Comment (mhils) [reply src/conversions/std/ipaddr.rs]: Done, thanks! 😃

**Review by adamreichold — APPROVED** [review]
The `extract` implementation is somewhat broad, but I think this is alright since the conversion are guided by the given types on the Rust side of things.

**Comment by adamreichold** [comment]
CI fails due to any unused import in the `sync` module.

**Comment by mhils** [comment]
> CI fails due to any unused import in the `sync` module.

Fixed, thanks! 🙇

**Comment by adamreichold** [comment]
@davidhewitt I would want to merge this but would be glad for a second opinion on the broadness of the `extract` implementation.

**Review by davidhewitt — APPROVED** [review]
Yep, this looks great to me - thanks both and sorry for the delay on my part!

File: src/conversions/std/ipaddr.rs
Comment (davidhewitt) [inline src/conversions/std/ipaddr.rs:10]: Actually, I suppose there is a good question about the width of the conversion here. I feel that normally we do some kind of type check on the way in, so this is a break from most of PyO3's convention?

Comment (davidhewitt) [reply src/conversions/std/ipaddr.rs:10]: Is the motivation not to check due to performance?

File: src/conversions/std/ipaddr.rs
Comment (davidhewitt) [inline src/conversions/std/ipaddr.rs:22]: Or is what's going on here that we're actually allowing strings as input? So this it's arguably convenient but this obviously doesn't round trip. Is that a problem?

File: src/conversions/std/ipaddr.rs
Comment (davidhewitt) [reply src/conversions/std/ipaddr.rs:10]: So not type-checking for exactly an ip address on the to-rust direction is consistent with what we've done for the recent `rust_decimal` integration, so I think this is fine?

Comment (adamreichold) [reply src/conversions/std/ipaddr.rs:10]: Yes, I think so too. But it was a point of contention in that review (of the `rust_decimal` integration) so I was uncomfortable just merging it without further discussion. As written before, I am personally completely fine with board `extract` impls (even parsing strings) as the user's Rust code controls the target types and the order in which they are applied.

Comment (mhils) [reply src/conversions/std/ipaddr.rs:10]: FWIW the entire `.packed` check is a performance optimization. Python's IP address classes internally store the address as an int. The fallback path (`obj.str()?.to_str()?.parse()`) would also work for `IPv4Address` and `IPv6Address` instances. I have no strong opinions here and would be happy to go with whatever you feel is most consistent. :)

From an API user perspective, the ability to pass in strings would be very nice as that matches existing Python APIs (`socket.socket` etc take strings, not IPvXAddress instances).

Comment (davidhewitt) [reply src/conversions/std/ipaddr.rs:10]: Yep, I think what we're converging on is that these generous-on-input conversions are quite practical.

At some point I wonder if we should go back over all the conversions we've implemented, categorise them into whether they are "generous" or strict, and make sure we're consistent in that approach?

**Comment by davidhewitt** [comment]
I think we're happy to see this move forward, so I'll click merge now. Thanks again!

File: src/conversions/std/ipaddr.rs
Comment (adamreichold) [reply src/conversions/std/ipaddr.rs:10]: > At some point I wonder if we should go back over all the conversions we've implemented, categorise them into whether they are "generous" or strict, and make sure we're consistent in that approach?

It does sound like a good idea to check them all for consistency. I just wonder when someone will the time to do that since there quite a few of them by now. Certainly worth making an issue to be resolve before 1.0.

Comment (davidhewitt) [reply src/conversions/std/ipaddr.rs:10]: Created #3226

**Comment by mhils** [comment]
Thank you two again! 🍰

It seems like the merge queue has hit an unrelated race condition, happy to fix things here if I'm misreading it. :)

**Comment by davidhewitt** [comment]
Yep no worries I will get #3225 merged and then retry.

**Pull request merged**
